File: solns/scala/27.scala
import scala.io.Source


def f(n: Int, a: Int, b: Int) = (n * n) + (a * n) + b

def count(ps: Set[Int], a: Int, b: Int): Int = {
  (0 until b).takeWhile(ps contains f(_, a, b)).size
}

def primes = Set() ++ Source.fromFile("../data/primes.txt")
    .getLines
    .map(_.trim.toInt)

val ps = primes.filter(_ < 1000)
val m = Map() ++ {
  for (
    a <- -1000 to 1000;
    b <- ps ++ ps.map(-1 * _);
    c =  count(ps, a, b)
  ) yield c -> (a, b)
}
val tuple = m(m.keys.max)
val result = tuple._1 * tuple._2

println(result)
assert(result == -59231)

File: solns/ruby/27.rb
require 'set'

# A simple quadratic
def f(n, a, b)
  n * n + a * n + b
end

# For how many values is a quadratic prime?
def count(ps, a, b)
  (0..b).take_while { |n| ps.include? f(n, a, b) }.count
end

# A list of the primes under 2 million will be fine
def primes
  file = File.new('../data/primes.txt')
  set = Set.new
  while (p = file.gets)
    set.add(p.to_i)
  end
  file.close
  set
end

b_max   = 1000
ps      = primes
bs      = ps.select { |b| b <= b_max }
bs += bs.map(&:-@)
a = 0
b = 0
c = 0

bs.each do |b_i|
  (-b_max..b_max).each do |a_i|
    c_i = count(ps, a_i, b_i)
    next unless c_i > c

    a = a_i
    b = b_i
    c = c_i
  end
end

result = a * b

puts result
raise Error unless result == -59_231
